source/extensions/filters/http/decompressor/decompressor_filter.cc

```diff
@@ -0,0 +1,221 @@
+                         *encoder_callbacks_, data, end_stream);
+}
+
+Http::FilterHeadersStatus DecompressorFilter::maybeInitDecompress(
+    DecompressorFilterConfig::DirectionConfig& direction_config,
+    Compression::Decompressor::DecompressorPtr& decompressor,
+    absl::variant<Http::StreamDecoderFilterCallbacks*, Http::StreamEncoderFilterCallbacks*>
+        decoder_or_encoder_callbacks,
+    Http::StreamFilterCallbacks& callbacks, Http::RequestOrResponseHeaderMap& headers) {
+  if (direction_config.decompressionEnabled() && !hasCacheControlNoTransform(headers) &&
+      contentEncodingMatches(headers)) {
+    direction_config.stats().decompressed_.inc();
+    decompressor = config_->makeDecompressor();
+    // remove the compression scheme's content encoding because the filter is going to decompress.
+    removeContentEncoding(headers);
+
+    // Per the filter documentation, only buffer if there is a Content-Length header present, and
+    // the filter is configured to buffer.
+    if (headers.removeContentLength() > 0 && direction_config.maxBufferedBytes()) {
+      direction_config.setBuffering(true);
+      if (absl::holds_alternative<Http::StreamDecoderFilterCallbacks*>(
+              decoder_or_encoder_callbacks)) {
+        absl::get<Http::StreamDecoderFilterCallbacks*>(decoder_or_encoder_callbacks)
+            ->setDecoderBufferLimit(direction_config.maxBufferedBytes().value());
+      } else {
+        absl::get<Http::StreamEncoderFilterCallbacks*>(decoder_or_encoder_callbacks)
+            ->setEncoderBufferLimit(direction_config.maxBufferedBytes().value());
+      }
+      direction_config.setHeaders(&headers);
+      // Note that the log will print the updated headers. The incoming headers can be seen from the
+      // log in decode/encodeHeaders.
+      ENVOY_STREAM_LOG(debug, "do decompress (without buffering) {}: {}", callbacks,
+                       direction_config.logString(), headers);
+      return Http::FilterHeadersStatus::StopIteration;
+    }
+    // FIX ME: chunked should not be added if we buffer, right?
+    sanitizeTransferEncoding(headers);
+
+    ENVOY_STREAM_LOG(debug, "do decompress (without buffering) {}: {}", callbacks,
+                     direction_config.logString(), headers);
+  } else {
+    ENVOY_STREAM_LOG(debug, "do not decompress {}: {}", callbacks, direction_config.logString(),
+                     headers);
+    direction_config.stats().not_decompressed_.inc();
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus
+DecompressorFilter::maybeDecompress(DecompressorFilterConfig::DirectionConfig& direction_config,
+                                    Compression::Decompressor::Decompressor* decompressor,
+                                    Http::StreamFilterCallbacks& callbacks,
+                                    Buffer::Instance& input_buffer, const bool end_stream) const {
+  if (decompressor) {
```

Comment:
I'd rather make `decompressor` an unique pointer and compare it with `nullptr` here.

Reply:
I could make it an optional ref if you'd rather not pass the raw pointer. However, I don't really want to move the ownership of the pointer into the function.